src/Pos/Wallet/Web/Server/Methods.hs

```diff
@@ -395,16 +395,22 @@ getWAddressBalance addr =
     getBalance <=< decodeCIdOrFail $ cwamId addr
 
 getWAddress :: WalletWebMode m => CWAddressMeta -> m CAddress
-getWAddress cAddr@CWAddressMeta {..} = do
+getWAddress cAddr = do
+    let aId = cwamId cAddr
     balance <- getWAddressBalance cAddr
+    let addrAccount = addrMetaToAccount cAddr
     (ctxs, _) <-
         searchHistory
             Nothing
-            (Just $ addrMetaToAccount cAddr)  -- just to specify addrId is not enough
-            (Just cwamId)
+            (Just addrAccount)  -- just to specify addrId is not enough
+            (Just aId)
             Nothing
     let isUsed = not (null ctxs) || balance > minBound
-    CAddress cwamId (mkCCoin balance) isUsed <$> isChangeAddress cAddr
+    -- Suppose we have transaction with inputs A1, A2, A3. Output address B_j is referred to as change address if it belongs to same account as one of A_i
+    acctAddrs <- map cwamId <$> getAccountAddrsOrThrow Ever addrAccount
+    let containsChangeAddr ctx = aId `elem` (ctOutputAddrs ctx) && not (null $ filter (`elem` acctAddrs) (ctInputAddrs ctx))
```

Comment:
`('elem' acctAddrs)` for each `ctInputAddrs ctx` will work for too long here, maybe put `actAddrs` to Set?